### ISSUE TITLE
the learning of rapid is inconsist when seeds in rapid is set to the same seeds

### ISSUE THREAD

**Issue by mao-xu** [daochenzha/rapid#2]
hi daochen,
When the seed in Rapid is set to 123, we run  5 experiments  on Minigrid-MultiRoom-N1-S10  and find  the training process and results are different for each group. Under the same parameters,  we only disable the rapid and run 5 experiments on Minigrid-MultiRoom-N1-S10. We find  the training process and results are same for each group.
Do we need to set other parameters during training rapid, in order to keep the same results when  the seed in Rapid is set to the same seed ?
Our environment is ubuntu18.04.5 lts，python 3.7.11， anaconda 4.10.1

**Comment by daochenzha** [comment]
@mao-xu Hi, I have just tried seed 123. It seems that the results are the same across runs in my side. I also used ubuntu. I would suggest manually seeding numpy, tf, and numpy module to see whether it works in your side.

**Comment by mao-xu** [comment]
Hi, @daochenzha, thank you for  your reply. According to your  suggestion, I have set the seed parameters before the main function runs， like     
     random.seed(123)
     os.environ['PYTHONHASHSEED'] = str(123)
     np.random.seed(123)
     tf.set_random_seed(123)
We run 2 experiments with 10000 steps on Minigrid-MultiRoom-N1-S10, and analyze the logs. The reward results is the same, but rapid_loss, policy_loss and value loss is not the same. The result likes
![C2AJ(8 L}~8@$TENUD Q$X](https://user-images.githubusercontent.com/4343768/137620277-ae04d9a7-d40e-492f-8140-f1af460899d7.png)
When I disable the rapid and only train the ppo, those are the same.
![ST2YOIUD%KYIQU5I1}Z10WF](https://user-images.githubusercontent.com/4343768/137620588-c6c6dacb-5cd2-4105-bd06-b076be3a4871.png)
Anything else I need to set up?

**Comment by daochenzha** [comment]
@mao-xu Thanks for bringing this up. I also observe similar results when running it agin. But unfortunately, I do not have a clue why it happens after checking the code. It seems that there is only one process and everything should be seeded.

**Comment by mao-xu** [comment]
Hi, @daochenzha, I also agree that the seed in rapid is set to the same seed. I modify the rapid and only train the  behavior cloning with the same input date. The rapid loss is a little bit different, I think the problem should be relevant to  the optimization of  the   sparse_softmax_cross_entropy_with_logits. Thank you for your reply.

**Comment by mao-xu** [comment]
Hi, @daochenzha , I modfiy the optimaization of the rapid_loss. At the same time, I test the modified rapid on  Minigrid-MultiRoom-N7-S4, and the results is the same. @daochenzha， you can test it，and the modified place is as follows：

        '''
        # RPID loss
        if isinstance(ac_space, Discrete):
            rapid_neglogpac = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=rapid_model.pi, labels=SL_A)
            rapid_loss = tf.reduce_mean(rapid_neglogpac)
        elif isinstance(ac_space, Box):
            rapid_loss = tf.reduce_mean(tf.square(rapid_model.pi-SL_A))
        rapid_grads = tf.gradients(rapid_loss, params)
        rapid_grads = list(zip(rapid_grads, params))
        '''
        # Trainer
        trainer = tf.train.AdamOptimizer(learning_rate=LR, epsilon=1e-5)
        _train = trainer.apply_gradients(grads)
       # _rapid_train = trainer.apply_gradients(rapid_grads)
       trainer_2 = tf.train.AdamOptimizer(learning_rate=LR, epsilon=1e-5)
       if isinstance(ac_space, Discrete):
            rapid_neglogpac = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=rapid_model.pi, labels=SL_A)
            rapid_loss = tf.reduce_mean(rapid_neglogpac)
       elif isinstance(ac_space, Box):
            rapid_loss = tf.reduce_mean(tf.square(rapid_model.pi - SL_A))

     #   rapid_grads = trainer_2.compute_gradients(rapid_loss, params)
     #   rapid_grads = list(zip(rapid_grads, params))
     #   _rapid_train = trainer_2.apply_gradients(rapid_grads)
        _rapid_train = trainer_2.minimize(rapid_loss)

**Comment by daochenzha** [comment]
@mao-xu Thanks for the fix! It is surprising to me that this will cause a difference. I will keep the code as it is since it is the original experimental code. I will all keep this issue open so that others who need to fix seed can see this.
